src/dom/progress-watcher.ts

```diff
@@ -0,0 +1,165 @@
+
+import { is } from '../is/is';
+import { mathf } from '../mathf/mathf';
+
+export interface ProgressWatcherItem {
+    range: number | Array<number>;
+    callback: Function;
+    inactiveCallback?: Function;
+    runWhen?: Function;
+}
+
+
+
+/**
+ * A class that provides callbacks for progress points.
+ *
+ * ```
+ * const progressWatcher = new ProgressWatcher();
+ *
+ *
+ * // Triggers continously while runs from 0.2, 0.4
+ *  this.progressWatcher.add({
+ *    range: [0.2, 0.4],
+ *    callback: (progress: number, direction: number)=> {
+ *          console.log(progress, direction);
+ *    }
+ *
+ *    // Runs when the progress is NOT 0.2 - 0.4
+ *    inactiveCallback: (progress: number, direction: number)=> {
+ *          console.log(progress, direction);
+ *    }
+ *  })
+ *
+ * // Triggers only when 0.2 is crossed in upwards or downwards direction.
+ *  this.progressWatcher.add({
+ *    range: 0.2,
+ *    callback: (progress: number, direction: number)=> {
+ *        // If going up
+ *        if(direction == -1) {
```

Comment:
Can you use an enum? https://www.typescriptlang.org/docs/handbook/enums.html

Reply:
Done.